test/__init__.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import collections
+import json
+import logging
+import os
+import tarfile
+import time
+
+import boto3
+import pytest
+import sagemaker
+import six
+
+import sagemaker_containers as smc
+
+DEFAULT_REGION = 'us-west-2'
+
+DEFAULT_CONFIG = dict(ContentType="application/x-numpy", TrainingInputMode="File",
+                      S3DistributionType="FullyReplicated", RecordWrapperType="None")
+
+DEFAULT_HYPERPARAMETERS = dict(sagemaker_region='us-west-2', sagemaker_job_name='sagemaker-training-job',
+                               sagemaker_enable_cloudwatch_metrics=False, sagemaker_container_log_level=logging.WARNING)
+
+
+def sagemaker_session(region_name=DEFAULT_REGION):  # type: (str) -> sagemaker.Session
+    return sagemaker.Session(boto3.Session(region_name=region_name))
+
+
+def default_bucket(session=None):  # type: (sagemaker.Session) -> str
+    session = session or sagemaker_session()
+    return session.default_bucket()
+
+
+def write_json(obj, path):  # type: (object, str) -> None
+    """Serialize ``obj`` as JSON in the ``path`` file.
+
+    Args:
+        obj (object): Object to be serialized
+        path (str): Path to JSON file
+    """
+    with open(path, 'w') as f:
+        json.dump(obj, f)
+
+
+def prepare(user_module, hyperparameters, channels, current_host='algo-1', hosts=None):
```

Comment:
Most of the functionality in this file seems to be around preparing the filesystem for simulating the EASE environment. Could we use the local mode logic from the python SDK for this instead? Even if it's not usable right out of the box for this, if all of the pieces of logic exist in the Python SDK, it could be straightforward to refactor it to decouple the environment preparation from the docker execution.

Reply:
It is correct that the functionality is preparing EASE. The difference is that local mode does that inside a container while we do not have a container here.

My vision here would be: we refactor the python sdk and local mode to have atomic functions so we can use it either inside or outside the container. I did not try to cover that in this PR because the size of the effort, although I agree that we need to review the python sdk design.

I tried to write a clean, isolated design, that can be used later to improve the sdk side.